### DISCUSSION TITLE
LogMonitor with named provider without GUID?

### DISCUSSION THREAD

**Discussion by mloskot** [microsoft/windows-container-tools#83] [Ideas]
This is a feature request to allow access to log equivalent to this

```
Get-WinEvent -ProviderName "My.Custom.Service.1.2.3"
```

I have a custom Windows service and I can use that PowerShell cmdlet to access the logs.

However, if I feed the LogMonitor with the following bit in my `LogMonitorConfig.json`:

```json
{
  "type": "ETW",
  "eventFormatMultiLine": false,
  "providers": [
    {
      "providerName": "My.Custom.Service.1.2.3",
      "level": "Information"
    }
  ]
}
```

then it complains about the invalid provider:

```
[2022-07-29T14:24:36.000Z][LOGMONITOR] ERROR: Invalid providers.
```

I read the source code and I see it is probing every named provider for GUID

https://github.com/microsoft/windows-container-tools/blob/83bcc982360d8041456ca0569d7cfccb4505c5b1/LogMonitor/src/LogMonitor/EtwMonitor.cpp#L129-L155

It would be nice to allow logs access by named providers only, without GUID.

**Comment by profnandaa** [comment]
@mloskot -- what's the reason around not having a GUID for a custom ETW provider?

**Comment by mloskot** [comment]
As in my example above, I install a simple custom Windows service (no installer, no GUIDs, etc.) and I can access its logs with this very simple command:
```
Get-WinEvent -ProviderName "My.Custom.Service.1.2.3"
```

Why not to make LogMonitor offer similar simple access, I'm wondering.

**Comment by profnandaa** [comment]
Dug through the code further and found out that the GUID is required by the API in `evntrace.h`, it's not optional.

```c
#if (WINVER >= _WIN32_WINNT_VISTA)
EXTERN_C
ULONG
WMIAPI
EnableTraceEx (
    _In_ LPCGUID ProviderId,
    _In_opt_ LPCGUID SourceId,
    _In_ TRACEHANDLE TraceHandle,
    _In_ ULONG IsEnabled,
    _In_ UCHAR Level,
    _In_ ULONGLONG MatchAnyKeyword,
    _In_ ULONGLONG MatchAllKeyword,
    _In_ ULONG EnableProperty,
    _In_opt_ PEVENT_FILTER_DESCRIPTOR EnableFilterDesc
    );
#endif
```

// ignore my previous comments
// following up on `providerWithoutGuid` map in code.

/cc. @iankingori

**Comment by profnandaa** [comment]
QQ, is your custom provider listed here when you run:
```ps
Get-WinEvent -ListProvider *
```

I'm also assuming that your initial `Get-WinEvent` command, you are running it within your container right?

**Comment by mloskot** [comment]
Yes, it is listed, like this:

```
Name     : My.Custom.Service.1.2.3
LogLinks : {Application}
Opcodes  : {}
Tasks    : {}
```

Yes, within container

**Comment by bobsira** [comment]
@mloskot The Get-WinEvent cmdlet gets events from event logs, including classic logs, such as the System and Application logs. The cmdlet gets data from event logs that are generated by the Windows Event Log technology and events in log files generated by Event Tracing for Windows (ETW).  
Get-WinEvent reads from ETW from a file and not from a provider name.  [Get-WinEvent source](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.diagnostics/get-winevent?view=powershell-7.2#example-14-event-tracing-for-windows:~:text=events%20in%20log%20files%20generated%20by%20Event%20Tracing%20for%20Windows%20(ETW))

Side notes on why we require GUID for ETW events take a look at this guide [ETW](https://docs.microsoft.com/en-us/windows-hardware/test/weg/instrumenting-your-code-with-etw)

Unless you have any more concerns, we will go ahead and close this.

cc @iankingori

**Comment by kevpar** [comment]
As an aside, there is a standard transformation that many ETW tools do to convert a name into a GUID. This allows users to specify a friendly name and have the conversion to GUID be consistent across many tools. While I don't think this would specifically solve @mloskot's problem, it could still be useful to support.

Details: https://docs.microsoft.com/en-us/archive/blogs/dcook/etw-provider-names-and-guids
Sample conversion code (Go): https://github.com/microsoft/go-winio/blob/main/pkg/etw/provider.go#L97-L127

**Comment by mloskot** [comment]
@bobsira Thank you for the explanation. I don't have anything to add here and, sadly, nothing to contribute. You're an expert here and I take your points, so please feel free to close it.

**Comment by bobsira** [comment]
@mloskot @kevpar 

An update is that we already have the feature of monitoring for logs without specifying the GUID as long as you have a valid provider name, and the provider is also a registered manifest-based event. Example configuration below would work fine: 

```
{
  "LogConfig": {
    "sources": [
      {
        "type": "ETW",
        "eventFormatMultiLine": false,
        "providers": [
          {
            "providerName": "Microsoft-Windows-WLAN-Drive",
            "level": "Information"
          }
        ]
      }
    ]
  }
}
```


[Code to the logic implementation](https://github.com/microsoft/windows-container-tools/blob/main/LogMonitor/src/LogMonitor/EtwMonitor.cpp#:~:text=//-,for%20(DWORD%20i%20%3D%200%3B%20i%20%3C%20penum%2D%3ENumberOfProviders%3B%20i%2B%2B),%7D,-//)

We'll update our docs to show it is possible to monitor ETW logs with only Provider Name and without the GUID.

The reason you were getting invalid providers error is because LogMonitor could not find a GUID associated with the provider's name you specified when it was looping through the list of provider names and their GUIDs in the system. 

These can probably be classified as event logs and not ETW logs.

**Comment by mloskot** [comment]
Thank you @bobsira and your team

**Comment by lucillex** [comment]
Hi @mloskot, I'm a PM on the Windows Container team working on Log Monitor.

Our team wanted to get some feedback on how you are using Log Monitor. We've started an issue [here](https://github.com/microsoft/windows-container-tools/issues/84). If you have any ideas for future features or general feedback, please provide them!

Thanks!
